### ISSUE TITLE
undefined should not throw error in Arraylike object.

### ISSUE THREAD

**Issue by ChristophAtAdlos** [mobxjs/serializr#58]
If a class has an uninitialized Array property, the object should be able to de-/serialize.

```
class Test {
    public myProp: MyObj[];
}
createModelSchema(Test, {
    myProp: list(object(MyObj));
});
```
Expected behavior:
```
{
  "myProp": undefined
}
```
(... or absence of the property.)

But got error: 
`Uncaught Error: [serializr] expected array (like) object`

**Comment by chengjianhua** [comment]
I transformed your ts codes into javascript code as below:

```
import { createModelSchema, list, object, serialize, deserialize } from 'serializr';

var MyObj = /** @class */ (function () {
  function MyObj() {
  }
  return MyObj;
}());
var Test = /** @class */ (function () {
  function Test() {
  }
  return Test;
}());

createModelSchema(Test, {
  myProp: list(object(MyObj))
});

const testModel = deserialize(Test, {});
const json = serialize(testModel);
```

The program throw error in `const json = serialize(testModel)`;

`deserialize()` is fine, the `serialize()` failed.

@ChristophAtAdlos Thanks for the issue that you raised, I found this problem, it need to be discussed with others own this repo , wait a moment ~ 😃

@mweststrate I read the code in type `list`

the `serializer` of `list` returned described below:

```
serializer: function (ar) {
            invariant(ar && "length" in ar && "map" in ar, "expected array (like) object")
            return ar.map(propSchema.serializer)
        },
```

@mweststrate It seems that a property typed as a list in a object have the possibility to be `null` or `undefined`, so may this situation should be considered ?

If yes, I'll create a PR for this.

**Comment by alexggordon** [comment]
I think most likely the way it is is because of what serializr aims to be. I don't think serializr tries to be a library that sets "defaults" for things. I think it tries to be a 1 - 1 mapping between object you pass in, and the class instances that are created. 

Think of this in a different situation. Say there's just a property like `someRelationship = object(MyObj)`. But you pass in `{someRelationship: undefined}`. What should that be? Should the relationship be still be undefined? Should it be a instance of `new MyObj()` with no properties set?

Currently, that above example returns undefined, so I think that not raising an error when passing a non-array like object is probably more in-line with that functionality. 

The other possible question for this is if we want some sort of enforceable type system. Maybe something like strict mode?

**Comment by ChristophAtAdlos** [comment]
Yes, I agree it's best to not set default values. A property which is undefined should not be defined.
As long as no exception is thrown, I'm happy.

**Comment by mweststrate** [comment]
Personally I think `deserialize` should fail as well :-P.

Note that in typescript, `x: Type[]` is not allowed with strict null checks, unless that `constructor` is guaranteed to put a value in it. Otherwise `x: Type[] | undefined` should have been used.

**Comment by chengjianhua** [comment]
Think about another situation. I use `@serializable` to decorate part of my model fields, I only need part of them to be serialized to `json`. I want to update the data in database, then I request `PUT /api/xx` with data `json`. 

```
class User {
  @serializable nick;
  @serializable age;
  @serializable gender;
  @serializable(list(primitive())) hobbies;
}

const user = new User();

user.age = 22;
user.nick = 'Nick';

const userJson = serialize(user); // { age: 20, nick: 'Nick' } 
```

Now I want to request `PUT /api/user/:id` with `userJson` to update the user informations in database, it's fine so far.

Now I request `GET /api/user/:id` to get the user informations from database, then I `const user = deserialize(User, response.data);` that aims to create a `User` model which some properties should be filled by the response data and is able to fill other informations later.

But it fails now, I think this could be a normal use case, so what the concepts behind the back of serializr prevent it ?

**Comment by alexggordon** [comment]
Maybe I missed something, but that shouldn't fail, as you're not passing `{hobbies: undefined}` to the model to the model though.

**Comment by alexggordon** [comment]
I think I'd tend to agree with @mweststrate, that `deserialize` should fail as well. I think it's too complicated to get into the area of setting defaults. Even array get's a little messy, because technically we don't error on a mobx array, (even though it's a fake array), so it's already accepting an array like object. 

However, it is a little weird that it selectively errors on deserilization/serialization. We should probably iron those errors out.

**Comment by zeevl** [comment]
Respectfully I disagree, there are plenty of cases when we might want an array to be undefined. 

If you're reading this because you're running into this issue as well, one solution is a custom serializer --
```
function undefinableList(propSchema: PropSchema) {
  return {
    serializer: ar => (ar ? list(propSchema).serializer(ar) : undefined),
    deserializer: list(propSchema).deserializer
  };
}
```

and use it like so --
```
  @serializable(undefinableList(primitive()))
  values: string[] | undefined;
```

**Comment by AndrewMorsillo** [comment]
I am also running into this. I am migrating my existing app to mobx and mobx-persist. I have several data structures where an array field may be null or undefined, which causes this error to be thrown. IMO it makes perfect sense for an array property to be null or undefined. Some code may even depend on that behavior, ex `if(!myObj.myArr){....initialize the arr}`.

**Comment by alexggordon** [comment]
@zeevl or @AndrewMorsillo can you give an example where setting the array default to empty array doesn't make the code perform as expected? (notice the hobbies array below)

```
class User {
  @serializable nick;
  @serializable age;
  @serializable gender;
  @serializable(list(primitive())) hobbies = [];
}
```

**Comment by AndrewMorsillo** [comment]
I've abandoned using serializr and mobx-persist for now because there were too many difficulties in making it work with my existing data. I guess it's current behavior would be fine if you were designing a new app but it is fairly hostile to working with existing data. 

As I stated above there may be code in existing apps that relies on a property being undefined. I don't think that is uncommon or an anti-pattern and IMO it should work out of the box with serializr. The serialized object should reflect the state of the object and sometimes that means properties being null or undefined.
